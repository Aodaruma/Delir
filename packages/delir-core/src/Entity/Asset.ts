import * as uuid from 'uuid'

export default class Asset {
    public id: string
    public mimeType: string

    /** Asset file extension (without `.` prefix) */
    public fileType: string
    public name: string
    public path: string

    constructor() {
        this.id = uuid.v4()
    }
}
