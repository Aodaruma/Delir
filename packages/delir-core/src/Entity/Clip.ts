import * as uuid from 'uuid'

export default class Clip {
    public id: string
    public renderer: string
    public placedFrame: number
    public durationFrames: number

    constructor() {
        this.id = uuid.v4()
    }
}
