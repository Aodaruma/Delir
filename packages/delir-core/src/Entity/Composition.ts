import * as uuid from 'uuid'
import { Branded } from '../helper/Branded'
import { safeAssign } from '../helper/safeAssign'
import ColorRGB from '../Values/ColorRGB'
import { Layer } from './Layer'
import { Project } from './Project'

interface CompositionProps {
    id?: string
    name: string
    width: number
    height: number
    framerate: number
    durationFrames: number
    samplingRate: number
    audioChannels: number
    backgroundColor: ColorRGB
}

type CompositionId = Branded<string, 'Entity/Composition/Id'>

class Composition implements CompositionProps {
    public id: Composition.Id
    public name: string
    public width: number
    public height: number
    public framerate: number
    public durationFrames: number
    public samplingRate: number
    public audioChannels: number
    public backgroundColor: ColorRGB
    /**
     * Layers rendering from [0].
     * The one that follows will overwrite the previous ones
     */
    public layers: ReadonlyArray<Layer> = []

    public project: Project

    constructor(props: CompositionProps) {
        this.id = uuid.v4() as Composition.Id
        safeAssign<Composition>(this, props as CompositionProps & {
            id: Composition.Id
        })
    }

    public patch(props: Partial<CompositionProps>) {
        safeAssign(this, props)
    }

    public findLayer(layerId: string): Layer | null {
        return this.layers.find(layer => layer.id === layerId) || null
    }

    public addLayer(layer: Layer, index: number | null = null): void {
        if (index == null) {
            this.layers = [layer, ...this.layers]
            return
        }

        const layers = [...this.layers]
        layers.splice(index, 0, layer)
        this.layers = layers
    }

    public removeLayer(layerId: string): boolean {
        const beforeLength = this.layers.length
        this.layers = this.layers.filter(layer => layer.id !== layerId)
        return this.layers.length !== beforeLength
    }

    public moveLayerIndex(layerId: string, newIndex: number): boolean {
        const index = this.layers.findIndex(layer => layer.id === layerId)
        if (index === -1) return false

        const clone = [...this.layers]
        this.layers = clone.splice(newIndex, 0, ...clone.splice(index, 1))
        return true
    }
}

namespace Composition {
    export type Id = CompositionId
}

export { Composition }
