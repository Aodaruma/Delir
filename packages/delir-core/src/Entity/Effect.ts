import * as uuid from 'uuid'

export default class Effect {
    public id: string
    public processor: string

    constructor() {
        this.id = uuid.v4()
    }
}
