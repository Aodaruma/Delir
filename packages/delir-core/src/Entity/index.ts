export { Asset } from './Asset'
export { Clip } from './Clip'
export { Composition } from './Composition'
export { Effect } from './Effect'
export { Keyframe , KeyframeValueTypes} from './Keyframe'
export { Layer } from './Layer'
export { Project } from './Project'
