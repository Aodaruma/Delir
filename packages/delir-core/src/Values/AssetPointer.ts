export default class AssetPointer {
    constructor(
        public assetId: string = ''
    ) {}
}
