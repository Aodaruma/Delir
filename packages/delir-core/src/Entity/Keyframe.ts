import * as uuid from 'uuid'

import AssetPointer from '../Values/AssetPointer'
import ColorRGB from '../Values/color-rgb'
import ColorRGBA from '../Values/color-rgba'
import Expression from '../Values/Expression'

export type KeyframeValueTypes = number | boolean | string | ColorRGB | ColorRGBA | Expression | AssetPointer | null

export default class Keyframe<T extends KeyframeValueTypes = KeyframeValueTypes> {
    public id: string
    public value: T
    public frameOnClip: number
    public easeInParam: [number, number]
    public easeOutParam: [number, number]

    constructor() {
        this.id = uuid.v4()
    }
}
