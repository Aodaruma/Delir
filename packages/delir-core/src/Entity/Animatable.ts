import { Expression } from '../Values'
import { Keyframe } from './Keyframe'

export class Animatable {
    public keyframes: { [paramName: string]: ReadonlyArray<Keyframe> } = Object.create(null)
    public expressions: { [paramName: string]: Expression } = Object.create(null)

    public findKeyframe(keyframeId: Keyframe.Id): Keyframe | null {
        for (const paramName of Object.keys(this.keyframes)) {
            const keyframe = this.keyframes[paramName].find(keyframe => keyframe.id === keyframeId)
            if (keyframe) return keyframe
        }

        return null
    }

    public findKeyframeAtFrame(paramName: string, frame: number): Keyframe | null {
        const sequence = this.keyframes[paramName]
        if (!sequence) return null

        for (const kf of sequence) {
            if (kf.frameOnClip === frame) {
                return kf
            }
        }

        return null
    }

    public findActiveKeyframeAt(paramName: string, frame: number): Keyframe | null {
        const sequence = this.keyframes[paramName]
        if (!sequence) return null

        for (let idx = 0, l = sequence.length; idx < l; idx++) {
            const pkf = sequence[idx - 1]
            const ckf = sequence[idx]
            const nkf = sequence[idx + 1]

            if (ckf.frameOnClip === frame) {
                return ckf
            }

            // [0]  10  20
            //       ◇   ◇
            // -> use 10frame's value at frame 0
            if (!pkf && frame <= ckf.frameOnClip) {
                return ckf
            }

            //  0  [10]
            //  ◇   |
            // -> use 0frame's value at frame 10
            if (!nkf && frame >= ckf.frameOnClip) {
                return ckf
            }
        }

        return null
    }

    public addKeyframe(paramName: string, keyframe: Keyframe): void {
        const newSequence = [
            ...(this.keyframes[paramName] || []),
            keyframe
        ].sort((a, b) => a.frameOnClip - b.frameOnClip)

        this.keyframes[paramName] = newSequence
    }

    public removeKeyframe(paramName: string, keyframeId: Keyframe.Id): boolean {
        const sequence = this.keyframes[paramName]
        if (!sequence) return false

        const beforeLength = sequence.length
        this.keyframes[paramName] = sequence.filter(keyframe => keyframe.id !== keyframeId)
        return this.keyframes[paramName].length !== beforeLength
    }

    public findExpression(paramName: string): Expression | null {
        return this.expressions[paramName] || null
    }

    public setExpression(paramName: string, expression: Expression | null) {
        if (expression == null) {
            delete this.expressions[paramName]
            return
        }

        this.expressions[paramName] = expression
    }
}
