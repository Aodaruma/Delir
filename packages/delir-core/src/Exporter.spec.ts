import { Asset, Clip, Composition, Effect, Keyframe, Layer, Project } from './Entity'
import AssetPointer from './Values/AssetPointer'
import ColorRGB from './Values/ColorRGB'
import Expression from './Values/Expression'

import * as Exporter from './Exporter'

describe('Export', () => {
    let project: Project

    beforeEach(() => {
        project = Object.assign(new Project(), {
            assets: [ Object.assign(new Asset(), { id: 'uuid-asset' }) ],
            compositions: [
                Object.assign(new Composition(), {
                    id: 'uuid-composition',
                    backgroundColor: new ColorRGB(0, 0, 0),
                    layers: [
                        Object.assign(new Layer(), {
                            id: 'uuid-layer',
                            clips: [
                                Object.assign(new Clip(), {
                                    id: 'uuid-clip',
                                    keyframes: {
                                        param: [ Object.assign(new Keyframe(), {
                                            id: 'uuid-clip-kf-1',
                                            value: new ColorRGB(0, 0, 0)
                                        }) ],
                                    },
                                    expressions: {
                                        param: new Expression('javascript', '1'),
                                    },
                                    effects: [
                                        Object.assign(new Effect(), {
                                            id: 'uuid-effect',
                                            expressions: {
                                                param: new Expression('javascript', '1')
                                            },
                                            keyframes: {
                                                param: [ Object.assign(new Keyframe(), {
                                                    id: 'uuid-effect-kf-1',
                                                    value: new ColorRGB(0, 0, 0),
                                                }) ],
                                                param2: [ Object.assign(new Keyframe(), {
                                                    id: 'uuid-effect-kf-2',
                                                    value: new AssetPointer(),
                                                }) ],
                                            },
                                        }),
                                    ],
                                })
                            ],
                        }),
                    ],
                }),
            ],
        })
    })

    it('Should correct serialize / desrialize project', () => {
        const serialized = Exporter.serialize(project)
        expect(serialized).toMatchSnapshot()
        expect(Exporter.deserialize(serialized)).toEqual(project)
    })
})
