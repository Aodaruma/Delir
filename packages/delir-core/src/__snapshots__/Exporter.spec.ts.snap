// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export Should correct serialize / desrialize project 1`] = `
Object {
  "__type": "entity:Project",
  "__value": Object {
    "assets": Array [
      Object {
        "__type": "entity:Asset",
        "__value": Object {
          "fileType": "png",
          "id": "uuid-asset",
          "name": "mock.png",
          "path": "/mocked/mock.png",
        },
      },
    ],
    "compositions": Array [
      Object {
        "__type": "entity:Composition",
        "__value": Object {
          "audioChannels": 2,
          "backgroundColor": Object {
            "__type": "value:ColorRGB",
            "__value": Object {
              "_blue": 0,
              "_green": 0,
              "_red": 0,
            },
          },
          "durationFrames": 100,
          "framerate": 30,
          "height": 100,
          "id": "uuid-composition",
          "layers": Array [
            Object {
              "__type": "entity:Layer",
              "__value": Object {
                "clips": Array [
                  Object {
                    "__type": "entity:Clip",
                    "__value": Object {
                      "durationFrames": 100,
                      "effects": Array [
                        Object {
                          "__type": "entity:Effect",
                          "__value": Object {
                            "expressions": Object {
                              "param": Object {
                                "__type": "value:Expression",
                                "__value": Object {
                                  "code": "1",
                                  "language": "javascript",
                                },
                              },
                            },
                            "id": "uuid-effect",
                            "keyframes": Object {
                              "param": Array [
                                Object {
                                  "__type": "entity:Keyframe",
                                  "__value": Object {
                                    "easeInParam": Array [
                                      1,
                                      1,
                                    ],
                                    "easeOutParam": Array [
                                      0,
                                      0,
                                    ],
                                    "frameOnClip": 0,
                                    "id": "uuid-effect-kf-1",
                                    "value": Object {
                                      "__type": "value:ColorRGB",
                                      "__value": Object {
                                        "_blue": 0,
                                        "_green": 0,
                                        "_red": 0,
                                      },
                                    },
                                  },
                                },
                              ],
                              "param2": Array [
                                Object {
                                  "__type": "entity:Keyframe",
                                  "__value": Object {
                                    "easeInParam": Array [
                                      1,
                                      1,
                                    ],
                                    "easeOutParam": Array [
                                      0,
                                      0,
                                    ],
                                    "frameOnClip": 0,
                                    "id": "uuid-effect-kf-2",
                                    "value": Object {
                                      "__type": "value:AssetPointer",
                                      "__value": Object {
                                        "assetId": "",
                                      },
                                    },
                                  },
                                },
                              ],
                            },
                            "processor": "mock",
                            "referenceName": null,
                          },
                        },
                      ],
                      "expressions": Object {
                        "param": Object {
                          "__type": "value:Expression",
                          "__value": Object {
                            "code": "1",
                            "language": "javascript",
                          },
                        },
                      },
                      "id": "uuid-clip",
                      "keyframes": Object {
                        "param": Array [
                          Object {
                            "__type": "entity:Keyframe",
                            "__value": Object {
                              "easeInParam": Array [
                                1,
                                1,
                              ],
                              "easeOutParam": Array [
                                0,
                                0,
                              ],
                              "frameOnClip": 0,
                              "id": "uuid-clip-kf-1",
                              "value": Object {
                                "__type": "value:ColorRGB",
                                "__value": Object {
                                  "_blue": 0,
                                  "_green": 0,
                                  "_red": 0,
                                },
                              },
                            },
                          },
                        ],
                      },
                      "placedFrame": 0,
                      "renderer": "video",
                    },
                  },
                ],
                "id": "uuid-layer",
                "name": "mocked-layer",
              },
            },
          ],
          "name": "mocked-composition",
          "samplingRate": 48000,
          "width": 100,
        },
      },
    ],
    "formatVersion": "2017091401",
  },
}
`;
